pkg/net/libp2p/libp2ptransport.go

```diff
@@ -0,0 +1,283 @@
+/*
+Copyright IBM Corp. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package libp2p
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/peerstore"
+	"github.com/multiformats/go-multiaddr"
+	"google.golang.org/protobuf/proto"
+
+	"github.com/filecoin-project/mir/pkg/events"
+	"github.com/filecoin-project/mir/pkg/logging"
+	mirnet "github.com/filecoin-project/mir/pkg/net"
+	"github.com/filecoin-project/mir/pkg/pb/eventpb"
+	"github.com/filecoin-project/mir/pkg/pb/messagepb"
+	"github.com/filecoin-project/mir/pkg/types"
+)
+
+const (
+	ID = "/mir/0.0.1"
+)
+
+type TransportMessage struct {
+	Sender  string
+	Payload []byte
+}
+
+var _ mirnet.Transport = &Transport{}
+
+type Transport struct {
+	host              host.Host
+	ownID             types.NodeID
+	membership        map[types.NodeID]string
+	incomingMessages  chan *events.EventList
+	outboundStreamsMx sync.Mutex
+	outboundStreams   map[types.NodeID]network.Stream
+	logger            logging.Logger
+}
+
+func NewTransport(h host.Host, membership map[types.NodeID]string, ownID types.NodeID, logger logging.Logger) *Transport {
+	if logger == nil {
+		logger = logging.ConsoleErrorLogger
+	}
+	return &Transport{
+		incomingMessages: make(chan *events.EventList),
+		outboundStreams:  make(map[types.NodeID]network.Stream),
+		logger:           logger,
+		membership:       membership,
+		ownID:            ownID,
+		host:             h,
+	}
+}
+
+func (t *Transport) ImplementsModule() {}
+
+func (t *Transport) EventsOut() <-chan *events.EventList {
+	return t.incomingMessages
+}
+
+func (t *Transport) Start() error {
+	t.logger.Log(logging.LevelDebug, fmt.Sprintf("node %s handler starting on %v", t.ownID, t.host.Addrs()))
+	t.host.SetStreamHandler(ID, t.mirHandler)
+	return nil
+}
+
+func (t *Transport) Stop() {
+	t.logger.Log(logging.LevelDebug, "Stopping libp2p transport.")
+	defer t.logger.Log(logging.LevelDebug, "libp2p transport stopped.")
+
+	t.outboundStreamsMx.Lock()
+	defer t.outboundStreamsMx.Unlock()
+
+	for id, s := range t.outboundStreams {
```

Comment:
I guess this is redundant. IIRC `host.Close()` implicitly closes all the active streams, but this is fine (is a good sanity-check).